src/tools/perf/api/libperf.h

```diff
@@ -98,6 +98,29 @@ enum {
 };
 
 
+#define FOREACH_UCP_PERF_DAEMON_AM_ID \
+    /* Host to daemon communication */  \
+    X(UCP_PERF_DAEMON_AM_ID_HOST_INIT,     0) \
+    X(UCP_PERF_DAEMON_AM_ID_HOST_REQ,      1) \
+    X(UCP_PERF_DAEMON_AM_ID_HOST_SEND_ACK, 2) \
+    X(UCP_PERF_DAEMON_AM_ID_HOST_RECV_ACK, 3) \
+    X(UCP_PERF_DAEMON_AM_ID_HOST_FIN,      4) \
+\
+    /* Peer to peer daemon communication */ \
+    X(UCP_PERF_DAEMON_AM_ID_PEER_INIT,     5) \
+    X(UCP_PERF_DAEMON_AM_ID_PEER_OP,       6)
+
+
+typedef enum {
+#define X(ID, VALUE) ID = VALUE,
+    FOREACH_UCP_PERF_DAEMON_AM_ID
```

Comment:
can we reuse UCS_PP_FOREACH or UCS_PP_FOREACH_SEP ?

Reply:
```diff
diff --git a/src/tools/perf/api/libperf.h b/src/tools/perf/api/libperf.h
index f3150d3dd2..dfe4438586 100644
--- a/src/tools/perf/api/libperf.h
+++ b/src/tools/perf/api/libperf.h
@@ -98,23 +98,13 @@ enum {
 };


-#define FOREACH_UCP_PERF_DAEMON_AM_ID \
-    /* Host to daemon communication */  \
-    X(UCP_PERF_DAEMON_AM_ID_HOST_INIT,     0) \
-    X(UCP_PERF_DAEMON_AM_ID_HOST_REQ,      1) \
-    X(UCP_PERF_DAEMON_AM_ID_HOST_SEND_ACK, 2) \
-    X(UCP_PERF_DAEMON_AM_ID_HOST_RECV_ACK, 3) \
-    X(UCP_PERF_DAEMON_AM_ID_HOST_FIN,      4) \
-\
-    /* Peer to peer daemon communication */ \
-    X(UCP_PERF_DAEMON_AM_ID_PEER_INIT,     5) \
-    X(UCP_PERF_DAEMON_AM_ID_PEER_OP,       6)
-
+#define UCP_PERF_DAEMON_AM_IDS \
+    HOST_INIT, HOST_REQ, HOST_SEND_ACK, HOST_RECV_ACK, HOST_FIN, PEER_INIT, \
+            PEER_OP

 typedef enum {
-#define X(ID, VALUE) ID = VALUE,
-    FOREACH_UCP_PERF_DAEMON_AM_ID
-#undef X
+#define UCP_PERF_DAEMON_AM_ID_DEFINE(_, _id) UCP_PERF_DAEMON_AM_ID_##_id
+    UCS_PP_FOREACH_SEP(UCP_PERF_DAEMON_AM_ID_DEFINE, _, UCP_PERF_DAEMON_AM_IDS)
 } ucp_perf_daemon_am_id_t;


diff --git a/src/tools/perf/perftest_daemon.c b/src/tools/perf/perftest_daemon.c
index 594f34db3b..06873cb462 100644
--- a/src/tools/perf/perftest_daemon.c
+++ b/src/tools/perf/perftest_daemon.c
@@ -40,10 +40,12 @@ static volatile int terminated = 0;

 const char *ucp_perf_daemon_am_id_name(ucp_perf_daemon_am_id_t am_id)
 {
+#define UCP_PERF_DAEMON_AM_ID_STR(_, _id) \
+    case UCP_PERF_DAEMON_AM_ID_##_id: \
+        return #_id;
+
     switch (am_id) {
-#define X(ID, _) case ID: return #ID;
-        FOREACH_UCP_PERF_DAEMON_AM_ID
-#undef X
+    UCS_PP_FOREACH(UCP_PERF_DAEMON_AM_ID_STR, _, UCP_PERF_DAEMON_AM_IDS)
         default: return "UNKNOWN";
     }
 }
```

Reply:
Ok, this is exactly what I was talking about)
In this diff you removed the id to make it suitable for `UCS_PP_FOREACH`.
Ok, that works, but the idea of X-macro is exactly the opposite: it is to group several related items in a single place, that's the whole point.

For example, for each message type we know in advance whether to expect reply-ep or rndv. So we can extend our X-macro to enforce this behavior:
```diff
+/* NAME, ID, REPLY-EP, RNDV */                                                                                                                                                                                                    
 #define FOREACH_UCP_PERF_DAEMON_AM_ID \                                                                                                                                                                                           
     /* Host to daemon communication */  \                                                                                                                                                                                         
-    X(UCP_PERF_DAEMON_AM_ID_HOST_INIT,     0) \                                                                                                                                                                                   
-    X(UCP_PERF_DAEMON_AM_ID_HOST_REQ,      1) \                                                                                                                                                                                   
-    X(UCP_PERF_DAEMON_AM_ID_HOST_SEND_ACK, 2) \                                                                                                                                                                                   
-    X(UCP_PERF_DAEMON_AM_ID_HOST_RECV_ACK, 3) \                                                                                                                                                                                   
-    X(UCP_PERF_DAEMON_AM_ID_HOST_FIN,      4) \                                                                                                                                                                                   
+    X(UCP_PERF_DAEMON_AM_ID_HOST_INIT,     0, 1, 0) \                                                                                                                                                                             
+    X(UCP_PERF_DAEMON_AM_ID_HOST_REQ,      1, 0, 0) \                                                                                                                                                                             
+    X(UCP_PERF_DAEMON_AM_ID_HOST_SEND_ACK, 2, 0, 0) \                                                                                                                                                                             
+    X(UCP_PERF_DAEMON_AM_ID_HOST_RECV_ACK, 3, 0, 0) \                                                                                                                                                                             
+    X(UCP_PERF_DAEMON_AM_ID_HOST_FIN,      4, 0, 0) \                                                                                                                                                                             
 \                                                                                                                                                                                                                                 
     /* Peer to peer daemon communication */ \                                                                                                                                                                                     
-    X(UCP_PERF_DAEMON_AM_ID_PEER_INIT,     5) \                                                                                                                                                                                   
-    X(UCP_PERF_DAEMON_AM_ID_PEER_OP,       6)                                                                                                                                                                                     
+    X(UCP_PERF_DAEMON_AM_ID_PEER_INIT,     5, 1, 0) \                                                                                                                                                                             
+    X(UCP_PERF_DAEMON_AM_ID_PEER_OP,       6, 0, 1)                                                                                                                                                                               
                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                   
 typedef enum {                                                                                                                                                                                                                    
-#define X(ID, VALUE) ID = VALUE,                                                                                                                                                                                                  
+#define X(ID, VALUE, REPLY_EP, RNDV) ID = VALUE,                                                                                                                                                                                  
     FOREACH_UCP_PERF_DAEMON_AM_ID                                                                                                                                                                                                 
 #undef X                                                                                                                                                                                                                          
 } ucp_perf_daemon_am_id_t;                                                                                                                                                                                                        

diff --git a/src/tools/perf/perftest_daemon.c b/src/tools/perf/perftest_daemon.c                                                                                                                                                   
index 594f34db3..293fa1e25 100644                                                                                                                                                                                                  
--- a/src/tools/perf/perftest_daemon.c                                                                                                                                                                                             
+++ b/src/tools/perf/perftest_daemon.c                                                                                                                                                                                             
@@ -38,10 +38,25 @@ typedef struct ucp_perf_daemon_context_t {                                                                                                                                                                     
                                                                                                                                                                                                                                   
 static volatile int terminated = 0;                                                                                                                                                                                               
                                                                                                                                                                                                                                   
+static int ucp_perf_daemon_am_id_reply_ep[] = {                                                                                                                                                                                   
+#define X(ID, VALUE, REPLY_EP, RNDV) \                                                                                                                                                                                            
+    [ID] = REPLY_EP,                                                                                                                                                                                                              
+#undef X                                                                                                                                                                                                                          
+};                                                     
+                                                       
+static int ucp_perf_daemon_am_id_rndv[] = {
+#define X(ID, VALUE, REPLY_EP, RNDV) \
+    [ID] = RNDV,                                       
+#undef X                                               
+};                                                     
+                                                       
@@ -221,10 +236,11 @@ static void ucp_perf_daemon_create_peer_ep(ucp_perf_daemon_context_t *ctx,                 
                                                        
 static void ucp_perf_daemon_check_am_msg(const ucp_am_recv_param_t *param,                                      
                                          size_t header_length,                                                  
-                                         int check_reply_ep,                                                    
-                                         int check_rndv,                                                        
                                          ucp_perf_daemon_am_id_t am_id)                                         
 {                                                      
+    int check_rndv     = ucp_perf_daemon_am_id_rndv[am_id];                                                     
+    int check_reply_ep = ucp_perf_daemon_am_id_reply_ep[am_id];                                                 
+                                                       
     ucs_trace_data("message received: %s", ucp_perf_daemon_am_id_name(am_id));                                  
                                                        
     ucs_assertv(check_rndv == !!(param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV),                                
@@ -283,7 +299,7 @@ ucp_perf_daemon_host_init_handler(void *arg, const void *header,                             
     void *value;                                       
     uint16_t value_length;                             
                                                        
-    ucp_perf_daemon_check_am_msg(param, header_length, 1, 0,                                                    
+    ucp_perf_daemon_check_am_msg(param, header_length,
                                  UCP_PERF_DAEMON_AM_ID_HOST_INIT);                                              
                                                        
     /* Peer EP must not be initialized on receiving HOST_INIT message.                                          
@@ -321,7 +337,7 @@ ucp_perf_daemon_host_req_handler(void *arg, const void *header,                              
     ucp_perf_daemon_req_t *req     = data;
     ucs_status_t status;                               
                                                        
-    ucp_perf_daemon_check_am_msg(param, header_length, 0, 0,                                                    
+    ucp_perf_daemon_check_am_msg(param, header_length,
                                  UCP_PERF_DAEMON_AM_ID_HOST_REQ);                                               
     ucs_assertv(length >= sizeof(*req), "length=%lu", length);                                                  
     ucs_assert(ctx->peer_ep != NULL);
```
So the advantage of X-macro is that you can easily extend your "enum" with more related fields. That's why using x-macro in enum declaration is so widely used practise.

Ok, that was the last attempt, if you really have strong opinion I will apply your diff

Reply:
Can you pls give example of places using X macro?
Also, are you planning to use X macro as described above (with additional fields) in subsequent PRs?

Reply:
Ok, after live discussion the agreement is to use X-macro flavour with explicit argument:
```
#define UCP_FOREACH_PERF_DAEMON_AM_ID(X) \
    X(UCP_PERF_DAEMON_AM_ID_HOST_INIT,     0) \
```